### PR TITLE
Allow disable apollo client cache

### LINKED ISSUE #4181: [security] Allow user disable apollo-client‘s local file cache
**Describe the solution you'd like**
add a `boolean localFileConfig` in apollo client, default value is true for compat.

**Additional context**
I am happy to work with this feature :)

### PR DESCRIPTION
**Pull request by Shoothzj**
## Which issue(s) this PR fixes:
Fixes #4181 

## Brief changelog
- add config `apollo.property.file.cache.enable`
- if `apollo.property.file.cache.enable` is false, use `RemoteConfigRepository`
- add corresponding unit tests

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Read the [Contributing Guide](https://github.com/ctripcorp/apollo/blob/master/CONTRIBUTING.md) before making this pull request.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit tests to verify the code.
- [x] Run `mvn clean test` to make sure this pull request doesn't break anything.
- [x] Update the [`CHANGES` log](https://github.com/ctripcorp/apollo/blob/master/CHANGES.md).

### CODE REVIEW

File: apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java [hunk 1]
```diff
@@ -56,7 +56,8 @@
   private boolean autoUpdateInjectedSpringProperties = true;
   private final RateLimiter warnLogRateLimiter;
   private boolean propertiesOrdered = false;
```
Comment (zouyx) [inline apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java:58]: why did you change the default value of  `propertyNamesCacheEnabled`? but did not change default value of `propertiesOrdered`?

File: apollo-client/src/main/java/com/ctrip/framework/apollo/spi/DefaultConfigFactory.java [hunk 2]
```diff
@@ -72,7 +72,9 @@ public Config create(String namespace) {
     // should not create a PropertiesCompatibleFileConfigRepository for them
     // calling the method `createLocalConfigRepository(...)` is more suitable
     // for ConfigFileFormat.Properties
-    if (ConfigFileFormat.isPropertiesCompatible(format) &&
+    if (!m_configUtil.isPropertyFileCacheEnabled()) {
```
Comment (nobodyiam) [inline apollo-client/src/main/java/com/ctrip/framework/apollo/spi/DefaultConfigFactory.java:75]: I think we could not add the logic here, as createConfigFile also will create local config repository.
How about we add another method, e.g. `createConfigRepository`, check whether the file cache enabled and create the correct config repository, then change all the places that call `createLocalConfigRepository` to `createConfigRepository`?

```java
  ConfigRepository createConfigRepository(String namespace) {
    if (m_configUtil.isPropertyFileCacheEnabled()) {
      return createLocalConfigRepository(namespace);
    }
    return createRemoteConfigRepository(namespace);
  }
```

File: apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java [hunk 3]
```diff
@@ -410,23 +412,38 @@ public boolean isPropertyNamesCacheEnabled() {
...
+            "false");
+  }
+
+  private void initPropertyFileCacheEnabled() {
+    propertyFileCacheEnabled = getPropertyBoolean(ApolloClientSystemConsts.APOLLO_PROPERTY_FILE_CACHE_ENABLE,
+            ApolloClientSystemConsts.APOLLO_PROPERTY_FILE_CACHE_ENABLE_ENVIRONMENT_VARIABLES,
+            "true");
+  }
+
+  private static boolean getPropertyBoolean(String propertyName, String envName, String defaultVal) {
```
Comment (nobodyiam) [inline apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java:431]: How about we keep it non-static?

```suggestion
  private boolean getPropertyBoolean(String propertyName, String envName, String defaultVal) {
```

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/ApolloClientSystemConsts.java [hunk 4]
```diff
@@ -146,4 +146,14 @@
...
+
+  /**
+   * enable property names cache
+   */
+  public static final String APOLLO_PROPERTY_FILE_CACHE_ENABLE = "apollo.property.file.cache.enable";
+
+  /**
+   * enable property names cache environment variables
+   */
+  public static final String APOLLO_PROPERTY_FILE_CACHE_ENABLE_ENVIRONMENT_VARIABLES = "APOLLO_PROPERTY_FILE_CACHE_ENABLE";
```
Comment (nobodyiam) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/ApolloClientSystemConsts.java:158]: How about simply apollo.file.cache.enable?
```suggestion

  /**
   * enable property names cache
   */
  public static final String APOLLO_FILE_CACHE_ENABLE = "apollo.file.cache.enable";

  /**
   * enable property names cache environment variables
   */
  public static final String APOLLO_FILE_CACHE_ENABLE_ENVIRONMENT_VARIABLES = "APOLLO_FILE_CACHE_ENABLE";
```

File: apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java (hunk 1 above)
Comment (Shoothzj) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java:58]: @zouyx I am not chaning the default value. I just extract a method for `propertyNamesCacheEnabled`

File: apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java (hunk 3 above)
Comment (Shoothzj) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java:431]: I rebase the code, fixed

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/ApolloClientSystemConsts.java (hunk 4 above)
Comment (Shoothzj) [reply apollo-core/src/main/java/com/ctrip/framework/apollo/core/ApolloClientSystemConsts.java:158]: I rebase the code, fixed

File: apollo-client/src/test/java/com/ctrip/framework/apollo/spi/DefaultConfigFactoryDisableFileCacheTest.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+public class DefaultConfigFactoryDisableFileCacheTest {
+
+    @Before
+    public void setUp() throws Exception {
+        MockInjector.setInstance(ConfigUtil.class, new MockConfigUtil());
+    }
+
+    @Test
+    public void testCreateFileDisableConfigRepository() throws Exception {
```
Comment (nobodyiam) [inline apollo-client/src/test/java/com/ctrip/framework/apollo/spi/DefaultConfigFactoryDisableFileCacheTest.java:35]: Add another test to verify the file cache enabled case?

File: apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java [hunk 6]
```diff
@@ -410,23 +412,38 @@ public boolean isPropertyNamesCacheEnabled() {
...
     if (Strings.isNullOrEmpty(enablePropertyNamesCache)) {
-      enablePropertyNamesCache = Foundation.app().getProperty(propertyName, "false");
+      enablePropertyNamesCache = Foundation.app().getProperty(propertyName, defaultVal);
     }
     if (!Strings.isNullOrEmpty(enablePropertyNamesCache)) {
       try {
-        propertyNamesCacheEnabled = Boolean.parseBoolean(enablePropertyNamesCache);
+        return Boolean.parseBoolean(enablePropertyNamesCache);
       } catch (Throwable ex) {
         logger.warn("Config for {} is invalid: {}, set default value: false",
```
Comment (nobodyiam) [inline apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java:443]: The `false` below should be replaced with the actual default value
```suggestion
        logger.warn("Config for {} is invalid: {}, set default value: {}",
```

File: apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java [hunk 7]
```diff
@@ -410,23 +412,38 @@ public boolean isPropertyNamesCacheEnabled() {
...
+            "false");
+  }
+
+  private void initPropertyFileCacheEnabled() {
+    propertyFileCacheEnabled = getPropertyBoolean(ApolloClientSystemConsts.APOLLO_FILE_CACHE_ENABLE,
+            ApolloClientSystemConsts.APOLLO_FILE_CACHE_ENABLE_ENVIRONMENT_VARIABLES,
+            "true");
+  }
+
+  private boolean getPropertyBoolean(String propertyName, String envName, String defaultVal) {
```
Comment (nobodyiam) [inline apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java:431]: the default value could simply be `boolean`?
```suggestion
  private boolean getPropertyBoolean(String propertyName, String envName, boolean defaultVal) {
```

File: apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java (hunk 1 above)
Comment (nobodyiam) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java:58]: The boolean variable's default value is false, so setting false or not is basically the same thing.
But this change does bring some confusion, so I would suggest we don't change it.

File: apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java (hunk 7 above)
Comment (Shoothzj) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java:431]: call `Foundation.app().getProperty()` must have a `String defaultVal`. I can change this to `boolean`, but it will looks ugly for doing many `String` `Boolean` convert. How about support `getBooleanProperty` in `Foundation.app()`? Or get without `String defaultVal`

**Comment by Shoothzj** [comment]
@nobodyiam After this change, the default value is seted in 
```java
  private void initPropertyNamesCacheEnabled() {
    propertyNamesCacheEnabled = getPropertyBoolean(ApolloClientSystemConsts.APOLLO_PROPERTY_NAMES_CACHE_ENABLE,
            ApolloClientSystemConsts.APOLLO_PROPERTY_NAMES_CACHE_ENABLE_ENVIRONMENT_VARIABLES,
            "false");
  }
```
If we keep the original, it might be confused for there are more than one code setting the init value.

File: apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java (hunk 6 above)
Comment (Shoothzj) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java:443]: fixed

File: apollo-client/src/test/java/com/ctrip/framework/apollo/spi/DefaultConfigFactoryDisableFileCacheTest.java (hunk 5 above)
Comment (Shoothzj) [reply apollo-client/src/test/java/com/ctrip/framework/apollo/spi/DefaultConfigFactoryDisableFileCacheTest.java:35]: added

**Comment by Shoothzj** [comment]
@nobodyiam PTAL again, thanks

File: apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java (hunk 7 above)
Comment (Anilople) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java:431]: The return type of getPropertyBoolean is boolean, so the default value is better keep type same as it. that will let method more understandable. 

We can keep the complex inside the method getPropertyBoolean, i.e convert boolean to String like follow?

```java
private boolean getPropertyBoolean(String propertyName, String envName, boolean defaultVal) {
    // ...
    Foundation.app().getProperty(propertyName, String.valueOf(defaultVal));
    // ...
}
```

Comment (Shoothzj) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java:431]: Yes we can, But if support `getBooleanProperty` or `getProperty` without default value, it will looks more graceful

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/ApolloClientSystemConsts.java [hunk 8]
```diff
@@ -146,4 +146,14 @@
    * enable property names cache environment variables
    */
   public static final String APOLLO_PROPERTY_NAMES_CACHE_ENABLE_ENVIRONMENT_VARIABLES = "APOLLO_PROPERTY_NAMES_CACHE_ENABLE";
+
+  /**
+   * enable property names cache
+   */
+  public static final String APOLLO_FILE_CACHE_ENABLE = "apollo.file.cache.enable";
```
Comment (Anilople) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/ApolloClientSystemConsts.java:153]: Is it possiable exists more cache in client in the future? 

Maybe we can plan the orgnaization of key, let it more extensible.

* apollo.cache.file.enable
* apollo.cache.config-map.enable
* apollo.cache.redis.enable

Use `apollo.cache.file.enable` or other key to instead of `apollo.file.cache.enable`?

File: apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java (hunk 7 above)
Comment (nobodyiam) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java:431]: we could simply pass null as the default value to read from Foundation.app()
```java
Foundation.app().getProperty(propertyName, null);
```

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/ApolloClientSystemConsts.java (hunk 8 above)
Comment (nobodyiam) [reply apollo-core/src/main/java/com/ctrip/framework/apollo/core/ApolloClientSystemConsts.java:153]: I agree `apollo.cache.file.enable` is better considering future extensibility.

File: apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java (hunk 7 above)
Comment (Shoothzj) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/util/ConfigUtil.java]: fixed

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/ApolloClientSystemConsts.java (hunk 8 above)
Comment (Shoothzj) [reply apollo-core/src/main/java/com/ctrip/framework/apollo/core/ApolloClientSystemConsts.java]: fixed

**Review by nobodyiam — DISMISSED** [review]
lgtm

**Review by nobodyiam — APPROVED** [review]
LGTM
